# CS 109A/STAT 121A/AC 209A/CSCI E-109A: Homework 6
# Reg-Logistic Regression, ROC, and Data Imputation

**Harvard University**<br/>
**Fall 2017**<br/>
**Instructors**: Pavlos Protopapas, Kevin Rader, Rahul Dave, Margo Levine

---

### INSTRUCTIONS

- To submit your assignment follow the instructions given in canvas.
- Restart the kernel and run the whole notebook again before you submit. 
- Do not include your name(s) in the notebook if you are submitting as a group. 
- If you submit individually and you have worked with someone, please include the name of your [one] partner below. 

---

Your partner's name (if you submit separately):

Enrollment Status (109A, 121A, 209A, or E109A):

Import libraries:

In [40]:
import numpy as np
import pandas as pd
import matplotlib
import matplotlib.pyplot as plt
from sklearn.linear_model import LogisticRegressionCV
import sklearn.metrics as metrics
from sklearn.preprocessing import PolynomialFeatures
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score
from sklearn.metrics import accuracy_score

from sklearn.tree import export_graphviz
from IPython.display import Image
from IPython.display import display
%matplotlib inline

from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix

## Automated Breast Cancer Detection

In this homework, we will consider the problem of early breast cancer detection from X-ray images. Specifically, given a candidate region of interest (ROI) from an X-ray image of a patient's breast, the goal is to predict if the region corresponds to a malignant tumor (label 1) or is normal (label 0). The training and test data sets for this problem is provided in the file `hw6_dataset.csv`. Each row in these files corresponds to a ROI in a patient's X-ray, with columns 1-117 containing features computed using standard image processing algorithms. The last column contains the class label, and is based on a radiologist's opinion or a biopsy. This data was obtained from the KDD Cup 2008 challenge.

The data set contain a total of 69,098 candidate ROIs, of which only 409 are malignant, while the remaining are all normal. 

*Note*: be careful of reading/treating column names and row names in this data set.

## Question 1: Beyond Classification Accuracy


0.  Split the data set into a training set and a testing set.  The training set should be 75% of the original data set, and the testing set 25%.  Use `np.random.seed(9001)`.

1. Fit a logistic regression classifier to the training set and report the  accuracy of the classifier on the test set. You should use $L_2$ regularization in logistic regression, with the regularization parameter tuned using cross-validation. 
    1. How does the fitted model compare with a classifier that predicts 'normal' (label 0) on all patients? 
    2. Do you think the difference in the classification accuracies are large enough to declare logistic regression as a better classifier than the all 0's classifier? Why or why not?
    
For applications with imbalanced class labels, in this case when there are many more healthy subjects ($Y=0$) than those with cancer ($Y=1$), the classification accuracy may not be the best metric to evaluate a classifier's performance. As an alternative, we could analyze the confusion table for the classifier. 

<ol start="3">
<li> Compute the confusion table for both the fitted classifier and the classifier that predicts all 0's.</li>
<li> Using the entries of the confusion table compute the *true positive rate* and the *true negative rate* for the two classifiers. Explain what these evaluation metrics mean for the specific task of cancer detection. Based on the observed metrics, comment on whether the fitted model is better than the all 0's classifier.</li>
<li> What is the *false positive rate* of the fitted classifier, and how is it related to its true positive and true negative rate? Why is a classifier with high false positive rate undesirable for a cancer detection task?</li>
</ol>
*Hint:* You may use the `metrics.confusion_matrix` function to compute the confusion matrix for a classification model.

In [6]:
np.random.seed(9001)
df = pd.read_csv('hw6_dataset.csv')
df.head()

-1.439999999999999891e-01  -1.429999999999999882e-01  \
0                   -0.01100                      0.138   
1                    0.21200                     -0.313   
2                    0.21500                     -0.184   
3                    0.27900                     -0.197   
4                    0.00922                     -0.138   

   -1.160000000000000059e-01  -1.029999999999999943e-01  \
0                    -0.2230                    -0.1730   
1                     0.2660                     0.2320   
2                     0.0274                     0.0494   
3                     0.1270                     0.0973   
4                     0.1690                     0.1540   

   2.260000000000000064e-01  2.099999999999999922e-01  \
0                     0.188                     0.284   
1                    -1.190                    -1.150   
2                     0.443                     0.463   
3                    -0.213                    -0.150   
4                    -0.391                    -0.397   

   -9.799999999999999822e-01  -7.800000000000000266e-01  \
0                    -0.0522                     -0.256   
1                    -1.8100                     -1.560   
2                    -1.0500                     -0.941   
3                    -1.3200                     -0.994   
4                    -1.6900                     -1.450   

   -4.739999999999999769e-01  -4.470000000000000084e-01  \
0                      0.129                      0.427   
1                     -1.250                     -1.200   
2                     -0.531                     -0.394   
3                     -1.110                     -1.090   
4                     -0.546                     -0.527   

             ...             9.250000000000000444e-01  \
0            ...                               -0.593   
1            ...                               -0.816   
2            ...                                0.634   
3            ...                               -0.640   
4            ...                               -0.277   

   5.160000000000000142e-01  3.439999999999999725e-01  \
0                     0.452                   0.00785   
1                     1.570                   0.39400   
2                     0.111                   0.37100   
3                     0.485                   0.29500   
4                     0.699                   0.37100   

   9.060000000000000275e-01  -1.129999999999999893e+00  \
0                    -0.533                    -0.0789   
1                     1.340                    -1.1800   
2                     0.859                    -0.9930   
3                     0.403                    -1.1200   
4                     0.481                    -1.0600   

   -5.520000000000000462e-01  5.530000000000000471e-01  \
0                      0.705                     0.906   
1                     -2.700                    -0.926   
2                     -0.492                     0.363   
3                     -0.343                     0.468   
4                     -0.526                     0.550   

   -4.169999999999999818e-01  2.560000000000000053e-01  \
0                      0.216                   -0.0723   
1                     -2.650                   -0.0447   
2                      0.326                   -0.0528   
3                     -0.820                    0.4350   
4                     -0.284                    0.1550   

   0.000000000000000000e+00  
0                       0.0  
1                       0.0  
2                       0.0  
3                       0.0  
4                       0.0  

[5 rows x 118 columns]

In [16]:
X = df.iloc[:, :-1]
y = df.iloc[:, -1]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)
X_train.shape, y_train.shape, X_test.shape, y_test.shape

((51822, 117), (51822,), (17275, 117), (17275,))

In [20]:
reg_Cs = np.hstack((10.**np.arange(-5, 0), 10**np.arange(0, 6)))

logit = LogisticRegressionCV(reg_Cs, cv=5)
logit.fit(X_train, y_train)
print("Logistic regression classifier on test set has accuracy:", logit.score(X_test, y_test))

0.995253256151


In [39]:
y_test.values
test_zeros = (y_test.values == 0).sum()
test_ones = (y_test.values == 1).sum()
test_total = y_test.values.shape[0]
# test_zeros, test_ones, test_zeros+test_ones, test_total
all_zero_accuracy = test_zeros/test_total
print("A classifier that predicts 'normal' (label 0) on patients in the test set has accuracy:", all_zero_accuracy)

A classifier that predicts 'normal' (label 0) on all patients has accuracy: 0.994674384949


## Answer:
### 2.A. How does the fitted model compare with a classifier that predicts 'normal' (label 0) on all patients?
- Using logistic regression, we get accuracy score on test set 99.53%, whereas using a "zero for all" classifier, we get test score of 99.47%.
- We can see that these two scores are very close.

### 2.B. Do you think the difference in the classification accuracies are large enough to declare logistic regression as a better classifier than the all 0's classifier? Why or why not?
- No, the difference between the classification accuracies are not large at all. So we cannot declare logistic regression is a better classfier than the all 0's classfier, if we only look at the two scores.
- This is because the dataset contains more than 69000 candidate ROIs, of which only 409 are malignant, while the remaining are all normal. So the ratio of people who does not have breast cancer is extremely higher than those with malignant tumor. The class labels are very imbalanced.

In [42]:
print("Confusion matrix for the fitted logistic regression classifier:")
print(confusion_matrix(y_test,logit.predict(X_test)))

Confusion matrix for the fitted logistic regression classifier:
[[17181     2]
 [   80    12]]


In [49]:
all_zero_confusion = np.array([[test_zeros, 0], [test_ones, 0]])
print("Confusion matrix for the all-zero classifier:")
print(all_zero_confusion)

Confusion matrix for the all-zero classifier:
[[17183     0]
 [   92     0]]


## Answer:
### 4. Using the entries of the confusion table compute the true positive rate and the true negative rate for the two classifiers. Explain what these evaluation metrics mean for the specific task of cancer detection. Based on the observed metrics, comment on whether the fitted model is better than the all 0's classifier.

#### Logistic Regression Classifier:
- True Positive Rate = $\frac{TP}{TP+FN} = \frac{12}{12+80} = 13.04\%$
- True Negative Rate = $\frac{TN}{TN+FP} = \frac{17181}{17181+2} = 99.99\%$

#### All-zero Classifier:
- True Positive Rate = $\frac{TP}{TP+FN} = \frac{0}{92+0} = 0\%$
- True Negative Rate = $\frac{TN}{TN+FP} = \frac{17183}{17183+0} = 100\%$

- True positive rate measures the proportion of positives that are correctly identified as such. In this case, true positive rate means the percentage of people with breast cancer who are correctly identified as having the condition.
- True negative rate measures the proportion of negatives that are correctly identified as such. In this case, true negative rate means the percentage of healthy people who are correctly identified as not having breast cancer.

- In this medical test condition, higher true positive rate is more important, because then we will be less likely to miss people who has breast cancer by telling them everything is fine. 
- Since all-zero classifier tells everybody that they are healthy, it is very useless in this medical area. 
- Therefore, the fitted logistic regression model is better thant the all 0's classifier.

### 5. What is the false positive rate of the fitted classifier, and how is it related to its true positive and true negative rate? Why is a classifier with high false positive rate undesirable for a cancer detection task?


## Question 2: ROC Analysis

Another powerful diagnostic tool for class-imbalanced classification tasks is the Receiver Operating Characteristic (ROC) curve. Notice that the default logistic regression classifier in `sklearn` classifies a data point by thresholding the predicted class probability $\hat{P}(Y=1)$ at 0.5. By using a different threshold, we can adjust the trade-off between the true positive rate (TPR) and false positive rate (FPR) of the classifier. The ROC curve allows us to visualize this trade-off across all possible thresholds.


1. Display the ROC curve for the fitted classifier on the *test set*. In the same plot, also display the ROC curve for the all 0's classifier. How do the two curves compare?

2.  Compute the highest TPR that can be achieved by the classifier at each of the following FPR's, and the thresholds at which they are achieved. Based on your results, comment on how the threshold influences a classifier's FPR.
    - FPR = 0
    - FPR = 0.1
    - FPR = 0.5
    - FPR = 0.9
- Suppose a clinician told you that diagnosing a cancer patient as normal is *twice* as critical an error as diagnosing a normal patient as having cancer. Based on this information, what threshold would you recommend the clinician to use? What is the TPR and FPR of the classifier at this threshold? 

- Compute the area under the ROC curve (AUC) for both the fitted classifier and the all 0's classifier. How does the difference in the AUCs of the two classifiers compare with the difference between their classification accuracies in Question 1, Part 2(A)? 

*Hint:* You may use the `metrics.roc_curve` function to compute the ROC curve for a classification model and the `metrics.roc_auc_score` function to compute the AUC for the model.

## Question 3: Missing data

In this problem you are given a different data set, `hw6_dataset_missing.csv`, that is  similar to the one you used above (same column definitions and same conditions), however this data set contains missing values. 

*Note*: be careful of reading/treating column names and row names in this data set as well, it *may* be different than the first data set.


1. Remove all observations that contain and missing values, split the dataset into a 75-25 train-test split, and fit the regularized logistic regression as in Question 1 (use `LogisticRegressionCV` again to retune).  Report the overall classification rate and TPR in the test set.
2. Restart with a fresh copy of the data in `hw6_dataset_missing.csv` and impute the missing data via mean imputation.  Split the data 75-25 and fit the regularized logistic regression model.  Report the overall classification rate and TPR in the test set.  
3. Again restart with a fresh copy of the data in `hw6_dataset_missing.csv` and impute the missing data via a model-based imputation method. Once again split the data 75-25 and fit the regularized logistic regression model.  Report the overall classification rate and TPR in the test set.  
4. Compare the results in the 3 previous parts of this problem.  Prepare a paragraph (5-6 sentences) discussing the results, the computational complexity of the methods, and conjecture and explain why you get the results that you see.



## APCOMP209a - Homework Question


This problem walks you through the derivation of the **likelihood equations** for a generalized linear model (GLM). Suppose that the random component of the GLM is in the univariate natural exponential family, so that
$$f(y_i|\theta_i) = h(y_i) e^{y_i\theta_i - b(\theta_i)}$$
Define the individual log-likelihood for each observation $i$ as
$$l_i(\theta_i) \equiv \log f(y_i|\theta_i)$$
with linear predictor
$$\eta_i = x_i^T\beta = g(\mu_i)$$
for some link function $g$ and where $\mu_i=E(Y_i)$.

1. Use the above expressions to write a simplified expression for the log-likelihood $l(\theta)$ for the entire dataset, $y_1, \dots, y_n$.

2. Use the chain rule to express $\frac{\partial l_i}{\partial \beta_j}$ in terms of the derivatives of $l_i, \theta_i, \mu_i$, and $\eta_i$. (*Hint*: Think carefully about which variables are related to which, and in what way. For example, for which of the above variables do you know the derivative with respect to $\beta_j$?)

3. Compute the derivatives for $\frac{\partial l_i}{\partial \theta_i}$ and $\frac{\partial \eta_i}{\partial \beta_j}$.

4. Express $\mu_i$ in terms of $\theta_i$, and use this relationship to compute $\frac{\partial \theta_i}{\partial \mu_i}$. (\emph{Hint}: Recall the cumulant function of a natural exponential family, and assume that you can write $\partial f/\partial g = (\partial g / \partial f)^{-1}$.)

5. Express $\eta_i$ in terms of $\mu_i$. Using the same hint as the above, compute $\frac{\partial \mu_i}{\partial \eta_i}$.

6. Put all of the above parts together to write an expression for $\frac{\partial l}{\partial \beta_j}$. Use matrix notation to write this expression as
$$\nabla_{\beta} l(\beta) = XDV^{-1}(Y - \mu) = 0$$
That is, compute the matrices $D$ and $V$ such that this equation holds.

7. If we use the canonical link function, how do your answers to part (6) simplify?

8. Finally, compute the above likelihood equations in the case of logistic regression, and show that this is equivalent to the solution given in lecture.


## Answer:
### 1. Use the above expressions to write a simplified expression for the log-likelihood $l(\theta)$ for the entire dataset, $y_1, \dots, y_n$.

$l(\theta) = \sum_{i=1}^{n}l_i \theta_i
= \sum_{i=1}^{n}log(h(y_i)) + \sum_{i=1}^{n}[y_i \theta_i - b(\theta_i)]$

### 2. . Use the chain rule to express $\frac{\partial l_i}{\partial \beta_j}$ in terms of the derivatives of $l_i, \theta_i, \mu_i$, and $\eta_i$. (*Hint*: Think carefully about which variables are related to which, and in what way. For example, for which of the above variables do you know the derivative with respect to $\beta_j$?)

$\frac{\partial l_i}{\partial \beta_j}$ = $\frac{\partial l_i}{\partial \theta_i} \cdot
\frac{\partial \theta_i}{\partial \mu_i} \cdot
\frac{\partial \mu_i}{\partial \eta_i} \cdot
\frac{\partial \eta_i}{\partial \beta_j}
$

### 3. Compute the derivatives for $\frac{\partial l_i}{\partial \theta_i}$ and $\frac{\partial \eta_i}{\partial \beta_j}$.

$\frac{\partial l_i}{\partial \theta_i} = y_i - b'(\theta_i) = y_i - \mu_i$

Since $\eta_i = x_i^T\beta = \beta_1 x_{i1} + \dots + \beta_p x_{ip}$, we get: $\frac{\partial \eta_i}{\partial \beta_j} = x_{ij}$

### 4. Express $\mu_i$ in terms of $\theta_i$, and use this relationship to compute $\frac{\partial \theta_i}{\partial \mu_i}$. (Hint: Recall the cumulant function of a natural exponential family, and assume that you can write $\partial f/\partial g = (\partial g / \partial f)^{-1}$.)

$\mu_i = b'(\theta_i)$

$\frac{\partial \theta_i}{\partial \mu_i} = (\frac{\partial \mu_i}{\partial \theta_i})^{-1}
= b''(\theta_i)^{-1} = (Var(y_i|\theta_i))^{-1} = \frac{1}{Var(y_i|\theta_i)}
$

### 5. Express $\eta_i$ in terms of $\mu_i$. Using the same hint as the above, compute $\frac{\partial \mu_i}{\partial \eta_i}$.

$\eta_i = g(\mu_i)$

$\frac{\partial \mu_i}{\partial \eta_i} = (\frac{\partial \eta_i}{\partial \mu_i})^{-1}
= g'(\mu_i)^{-1} = \frac{1}{g'(\mu_i)}
$

### 6. Put all of the above parts together to write an expression for $\frac{\partial l}{\partial \beta_j}$. Use matrix notation to write this expression as
$$\nabla_{\beta} l(\beta) = XDV^{-1}(Y - \mu) = 0$$
That is, compute the matrices $D$ and $V$ such that this equation holds.

$\frac{\partial l_i}{\partial \beta_j} = \frac{\partial l_i}{\partial \theta_i} \cdot
\frac{\partial \theta_i}{\partial \mu_i} \cdot
\frac{\partial \mu_i}{\partial \eta_i} \cdot
\frac{\partial \eta_i}{\partial \beta_j}
= (y_i - \mu_i) \cdot \frac{1}{Var(y_i|\theta_i)} \cdot \frac{1}{g'(\mu_i)} \cdot x_{ij}
$

$\frac{\partial l}{\partial \beta_j}
= \sum_{i=1}^{n}(y_i - \mu_i) \cdot \frac{1}{Var(y_i|\theta_i)} \cdot \frac{1}{g'(\mu_i)} \cdot x_{ij}
$

In matrix form, y and $\mu$ are column vectors, and X is the data matrix.

If we write this expression using matrix notation as $\nabla_{\beta} l(\beta) = XDV^{-1}(Y - \mu) = 0$, we will get the following matrices $D$ and $V^{-1}$:

$D = 
\begin{bmatrix}
g'(\mu_1)^{-1} & 0 & \cdots & 0 \\
0 & g'(\mu_2)^{-1} & \cdots & 0 \\
\vdots & \vdots & \ddots & \vdots \\
0 & 0 & 0 & g'(\mu_n)^{-1}
\end{bmatrix}
$

$V =
\begin{bmatrix}
Var(y_1|\theta_1) & 0 & \cdots & 0 \\
0 & Var(y_2|\theta_2) & \cdots & 0 \\
\vdots & \vdots & \ddots & \vdots \\
0 & 0 & 0 & Var(y_n|\theta_n)
\end{bmatrix}
$



### 7. If we use the canonical link function, how do your answers to part (6) simplify?

Canonical link: $\theta = \eta$

Therefore, we have $\frac{\partial \theta_i}{\partial \mu_i} \cdot \frac{\partial \mu_i}{\partial \eta_i} = 1$

The answer to part (6) can be simplified as $\nabla_{\beta} l(\beta) = X(Y - \mu) = 0$

### 8. Finally, compute the above likelihood equations in the case of logistic regression, and show that this is equivalent to the solution given in lecture.